src/locale.js

```diff
@@ -76,25 +76,49 @@ export default class Locale {
-		 * `<index>` is the index in the `values` array.
+		 * This method's context is statically bound to the `Locale` instance and **always should be called as a function**:
 		 *
-		 *		editor.t( 'Created file "%0" in %1ms.', [ fileName, timeTaken ] );
+		 *		const t = locale.t;
+		 *		t( 'Label' );
 		 *
-		 * This method's context is statically bound to Locale instance,
-		 * so it can be called as a function:
+		 * The message can be either a string or an object implementing the {@link module:utils/translation-service~Message} interface.
 		 *
-		 *		const t = this.t;
-		 *		t( 'Label' );
+		 * The message may contain placeholders (`%<index>`) for value(s) that are passed as the second argument.
+		 * For an array of values the `%<index>` will be changed to an element of that array at the given index.
+		 * For a single value passed as the second argument, only the `%0` placeholders will be changed to the provided value.
+		 *
+		 *		t( 'Created file "%0" in %1ms.', [ fileName, timeTaken ] );
+		 * 		t( 'Created file "%0", fileName );
+		 *
+		 * The message supports plural forms. To specify a plural form, use the `plural` property. Single or plural form
+		 * will be chosen depending on the first value from the passed `values`. The value of this property is used
+		 * as a default plural translation when the translation for the target language is missing. Therefor, it should be a number.
+		 *
+		 *		t( { string: 'Add a space', plural: 'Add %0 spaces' }, 1 ); // 'Add a space' for the English language.
+		 *		t( { string: 'Add a space', plural: 'Add %0 spaces' }, 5 ); // 'Add 5 spaces' for the English language.
+		 *		t( { string: '%1 a space', plural: '%1 %0 spaces' }, [ 2, 'Add' ] ); // 'Add 2 spaces' for the English language.
+		 *
+		 * 		t( { string: 'Add a space', plural: 'Add %0 spaces' }, 1 ); // 'Dodaj spację' for the Polish language.
+		 *		t( { string: 'Add a space', plural: 'Add %0 spaces' }, 5 ); // 'Dodaj 5 spacji' for the Polish language.
+		 *		t( { string: '%1 a space', plural: '%1 %0 spaces' }, [ 2, 'Add' ] ); // 'Dodaj 2 spacje' for the Polish language.
+		 *
+		 * The message can provide a context using the `context` property when the message ids created from message strings
+		 * are not unique. When the `context` property is set the message id will be constructed in
+		 * the following way: `${ message.string }_${ message.context }`. This context will be also used
+		 * by translators later as an additional context for the translated message.
+		 *
+		 *		t( { string: 'image', context: 'Add/Remove image' } );
 		 *
 		 * @method #t
-		 * @param {String} str The string to translate.
-		 * @param {String[]} [values] Values that should be used to interpolate the string.
+		 * @param {String|module:utils/translation-service~Message} message A message that will be localized (translated).
+		 * @param {String|Number|Array.<String|Number>} [values] A value or an array of values that will fill message placeholders.
```

Comment:
argue with @scofalik :smile: 

For me the cost of writing `[]` around one argument is rather low. But it's true that in our use cases it's ~90% of all cases.

Reply:
I can see we didn't support this shorter form previously. But I agree that in great majority of cases there will be just one placeholder, so this may be a useful addition. OTOH, this function is used so few times that I'd personally do KISS, as @mlewand mentions.

Reply:
After some extra f2f talking we decided we're ok with the way it was proposed by @ma2ciek in this PR (so both primitive and array variants are accepted).